content/blog/posts/vectors-matrices-matrix-order.md

```diff
@@ -0,0 +1,732 @@
+const float rowVector[3] = {1.f, 0.f, 1.f};
+
+// 3x4 matrix in row-major order
+const std::size_t rows = 3;
+const std::size_t cols = 4;
+
+const float rowMajor[rows][cols] = 
+{ 
+    {1.f, 0.f, 1.f, 4.f}, // row 0 
+    {0.f, 2.f, 0.f, 1.f}, // row 1
+    {3.f, 0.f, 3.f, 4.f}  // row 2
+};
+
+// 1x4 row vector to hold the result
+float result[cols] = {0.f, 0.f, 0.f, 0.f};
+
+for(std::size_t c = 0; c < cols; c++)
+{
+   for(std::size_t r = 0; r < rows; r++)
+   {
+      result[c] += rowMajor[r][c] * rowVector[r];
+   }
+}
+
+// prints 4 0 4 8
+for(std::size_t i = 0; i < cols; i++)
+{
+   std::cout << result[i] << " ";
+}
+```
+
+Now consider the following product between the transpose of the above matrix and a column vector:
+
+$$
+\begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \\\ 4 & 1 & 4  \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} = \begin{bmatrix} 4 \\\ 0 \\\ 4 \\\ 8 \end{bmatrix}
+$$
+
+We can implement the matrix-vector product with column major array storage as follows:
+
+```c++
+// 3x1 column vector
+const float columnVector[3] = {1.f, 0.f, 1.f};
+
+// 4x3 matrix in column-major order
+const std::size_t rows = 4;
+const std::size_t cols = 3;
+
+const float columnMajor[cols][rows] = 
+{ 
+    {1.f, 0.f, 1.f, 4.f}, // column 0 
+    {0.f, 2.f, 0.f, 1.f}, // column 1
+    {3.f, 0.f, 3.f, 4.f}  // column 2
+};
+
+// 4x1 column vector to hold the result
+float result[rows] = {0.f, 0.f, 0.f, 0.f};
+
+for(std::size_t r = 0; r < rows; r++)
+{
+   for(std::size_t c = 0; c < cols; c++)
+   {
+      result[r] += columnMajor[c][r] * columnVector[c];
+   }
+}
+
+// prints 4 0 4 8
+for(std::size_t i = 0; i < rows; i++)
+{
+   std::cout << result[i] << " ";
+}
+```
+
+You may have noticed that the code for both snippets is identical, bar some flipping of variable names for clarity. Remember that the product of a row vector with a given matrix is the same as the product of a column vector of the transpose of that same matrix:
+
+$$
+\vec{r} \cdot \mathbf{M} =\mathbf{M}^{T} \cdot \vec{c} 
+$$
+
+An interesting property of storing our matrices in the array matrix order convention of our vector convention (i.e. row vectors with matrices stored in row-major order and column vectors with matrices stored in column-major order) is that the actual array data for the matrices is identical regardless of which of the two convention combinations you choose. Consider the following 2d transform matrix for scaling, anti-clockwise rotating and translating a given row vector:
+
+$$
+\begin{bmatrix}      S_{x} \cdot \cos(\theta) & \sin(\theta) & 0  \\\      -\sin(\theta) & S_{y} \cdot \cos(\theta) & 0  \\\  T_{x} &  T_{y}  & 1  \end{bmatrix}
+$$
+
+When laid out in row-major in a two dimensional array, the array is initialized as follows:
+
+```c++
+// Scale to double and quadruple the x and y axis accordingly
+const float Sx = 2.f;
+const float Sy = 4.f
+
+// Rotation of approx. 45 degrees (in radians)
+const float theta = 0.707;
```

Comment:
This looks like cos theta rather than theta. 45 degrees in radians would be about 0.785.

Reply:
```suggestion
const float theta = 0.785;
```